modules/fedimint-ecash-migration-server/src/lib.rs

```diff
@@ -169,54 +183,248 @@ impl ServerModuleInit for EcashMigrationInit {
+                .await;
+        }
+
+        Ok(())
     }
 
     async fn process_input<'a, 'b, 'c>(
         &'a self,
-        _dbtx: &mut DatabaseTransaction<'c>,
-        _input: &'b EcashMigrationInput,
+        dbtx: &mut DatabaseTransaction<'c>,
+        input: &'b EcashMigrationInput,
         _in_point: InPoint,
     ) -> Result<InputMeta, EcashMigrationInputError> {
-        Err(EcashMigrationInputError::NotSupported)
+        let (transfer_id, note, amount) = match input {
+            EcashMigrationInput::RedeemOriginEcash {
+                transfer_id,
+                note,
+                amount,
+            } => (transfer_id, note, amount),
+            EcashMigrationInput::Default { variant, .. } => {
+                return Err(EcashMigrationInputError::UnknownInputVariant(*variant));
+            }
+        };
+
+        // Check transfer is active and exists
+        if !is_transfer_active(dbtx, *transfer_id, self.num_peers).await {
+            return Err(EcashMigrationInputError::InvalidTransfer(*transfer_id));
+        }
+
+        // Get current balance in the transfer
+        let deposited_amount = dbtx
+            .get_value(&DepositedAmountKey(*transfer_id))
+            .await
+            .expect("Deposited amount not found for existing transfer");
+        let withdrawn_amount = dbtx
+            .get_value(&WithdrawnAmountKey(*transfer_id))
+            .await
+            .expect("Withdrawn amount not found for existing transfer");
+        let transfer_balance = deposited_amount
+            .checked_sub(withdrawn_amount)
+            .expect("Liability transfer balance cannot be negative");
+
+        // Check amount is within transfer balance
+        if transfer_balance < *amount {
+            return Err(EcashMigrationInputError::UnderfundedTransfer);
+        }
+
+        // Check note signature
+        let denomination_key = dbtx
+            .get_value(&DenominationKeyKey {
+                transfer_id: *transfer_id,
+                amount: *amount,
+            })
+            .await
+            .ok_or(EcashMigrationInputError::InvalidAmountTier(*amount))?;
+        if !note.verify(denomination_key) {
+            return Err(EcashMigrationInputError::InvalidSignature);
+        }
+
+        // Check if already redeemed on either the origin or local spend book
+        let is_spent_on_origin = dbtx
+            .get_value(&OriginSpendBookKey {
+                transfer_id: *transfer_id,
+                nonce: note.nonce,
+            })
+            .await
+            .is_some();
+        let is_spent_on_local = dbtx
+            .get_value(&LocalSpendBookKey {
+                transfer_id: *transfer_id,
+                nonce: note.nonce,
+            })
+            .await
+            .is_some();
+        if is_spent_on_origin || is_spent_on_local {
+            return Err(EcashMigrationInputError::AlreadyRedeemed(note.nonce));
```

Comment:
Would split this error, because there's an important difference and caller might want to know which one was it.